Refactoring gitwash (#23588)
matplotlib/matplotlib: doc/devel/development_workflow.rst

Development workflow
####################

Workflow summary
================

In what follows we'll refer to the upstream Matplotlib ``main`` branch, as
"trunk".

* Don't use your ``main`` branch for anything.  Consider deleting it.
* When you are starting a new set of changes, fetch any changes from ``main``,
  and start a new *feature branch* from that.
* Make a new branch for each separable set of changes — "one task, one
  branch".
* Name your branch for the purpose of the changes - e.g.
  ``bugfix-for-issue-14`` or ``refactor-database-code``.

> * If you get stuck, reach out on Gitter or the Mailing list.

Reviewer: ```suggestion
* If you get stuck, reach out on Gitter or `discourse <https://discourse.matplotlib.org>`__.
```

Development workflow
####################

Workflow summary
================

In what follows we'll refer to the upstream Matplotlib ``main`` branch, as

> "trunk".

Reviewer 2: I think the term *trunk* was quite common in SVN, but is not in git. I suggest to use `main` or `upstream/main` and not introduce a new name.

* Don't use your ``main`` branch for anything.  Consider deleting it.

> * When you are starting a new set of changes, fetch any changes from ``main``,

Reviewer 2: ```suggestion
* Before starting a new set of changes, fetch all changes from ``upstream/main``,
```

  and start a new *feature branch* from that.

> * Make a new branch for each separable set of changes — "one task, one
>   branch".

Reviewer 2: Better be concrete?
```suggestion
* Make a new branch for each feature or bugfix — "one task, one
  branch".
```

* Name your branch for the purpose of the changes - e.g.
  ``bugfix-for-issue-14`` or ``refactor-database-code``.
* If you get stuck, reach out on Gitter or
  `discourse <https://discourse.matplotlib.org>`__.

> * Ask for a code review!

Reviewer 2: ```suggestion
* When your ready or need feedback on your code, open a pull-request
  so that the matplotlib developers can give feedback and eventually
  include your suggested code into the ``main`` branch.
```

Development workflow
####################

Workflow summary
================

To keep your work well organized, with readable history, and in turn make it
easier for project maintainers (that might be you) to see what you've done, and
why you did it, we recommend the following:

* Don't use your ``main`` branch for anything.  Consider deleting it.
* Before starting a new set of changes, fetch all changes from
  ``upstream/main``, and start a new *feature branch* from that.
* Make a new branch for each feature or bug fix — "one task, one branch".
* Name your branch for the purpose of the changes - e.g.
  ``bugfix-for-issue-14`` or ``refactor-database-code``.
* If you get stuck, reach out on Gitter or
  `discourse <https://discourse.matplotlib.org>`__.

> * When your ready or need feedback on your code, open a pull-request so that the

Reviewer 3: ```suggestion
* When you're ready or need feedback on your code, open a pull-request so that the
```

  Matplotlib developers can give feedback and eventually include your suggested
  code into the ``main`` branch.

.. note::

   It may sound strange, but deleting your own ``main`` branch can help reduce
   confusion about which branch you are on.  See `deleting main on github`_ for
   details.

.. _deleting main on github: https://matthew-brett.github.io/pydagogue/gh_delete_master.html

.. _update-mirror-main:

Update the mirror of main
=========================

First make sure you have done :ref:`linking-to-upstream`.

> From time to time you should fetch the upstream changes from github::

Reviewer 3: ```suggestion
From time to time you should fetch the upstream changes from GitHub::
```

   ``commit``. The ``m`` flag just signals that you're going to type a
   message on the command line.  The ``a`` flag — you can just take on
   faith — or see `why the -a flag?`_ — and the helpful use-case
   description in the `tangled working copy problem`_. The
   `git commit <https://git-scm.com/docs/git-commit>`_ manual page might also be
   useful.
#. To push the changes up to your forked repo on github, do a ``git
   push``.

.. _why the -a flag?: http://gitready.com/beginner/2009/01/18/the-staging-area.html
.. _tangled working copy problem: http://2ndscale.com/rtomayko/2008/the-thing-about-git

Ask for your changes to be reviewed or merged
=============================================

When you are ready to ask for someone to review your code and consider a merge,
`submit your Pull Request (PR) <https://docs.github.com/pull-requests>`_.

Enter a title for the set of changes, and some explanation of what you've done.
Say if there is anything you'd like particular attention for - like a
complicated change or some code you are not happy with.

If you don't think your request is ready to be merged, just say so in your pull

> request message.  This is still a good way of getting some preliminary code

Reviewer: ```suggestion
request message and use the "Draft PR" feature of GitHub.  
This is a good way of getting some preliminary code
```

     #	INSTALL
     no changes added to commit (use "git add" and/or "git commit -a")

#. Check what the actual changes are with ``git diff``.
#. Add any new files to version control ``git add new_file_name``.
#. To commit all modified files into the local copy of your repo,, do
   ``git commit -am 'A commit message'``.  Note the ``-am`` options to
   ``commit``. The ``m`` flag just signals that you're going to type a
   message on the command line.  The ``a`` flag — you can just take on
   faith — or see `why the -a flag?`_ — and the helpful use-case
   description in the `tangled working copy problem`_. The
   `git commit <https://git-scm.com/docs/git-commit>`_ manual page might also be
   useful.
#. To push the changes up to your forked repo on github, do a ``git
   push``.

.. _why the -a flag?: http://gitready.com/beginner/2009/01/18/the-staging-area.html
.. _tangled working copy problem: http://2ndscale.com/rtomayko/2008/the-thing-about-git

> Ask for your changes to be reviewed or merged
> =============================================

Reviewer 2: ```suggestion
Open a Pull Request
===================
```

   git log

> You can also look at the `network graph visualizer <https://github.com/matplotlib/matplotlib/network>`_ for your github

Reviewer: We should remove this, the view for Matplotlib is approximately useless (we have in the past asked GitHub ahout it and their response was "ahhh, your network is too big....")

   git log

.. _rebase-on-trunk:

> Rebasing on trunk

Reviewer 2: ```suggestion
Rebasing on ``upstream/main``
```

See `rebase without tears`_ for more detail.

.. _rebase without tears: https://matthew-brett.github.io/pydagogue/rebase_without_tears.html

> To do a rebase on trunk::
> 
>     # Update the mirror of trunk

Reviewer 2: ```suggestion
To do a rebase on ``upstream/main``::

    # Fetch changes from upstream/main
```
